### ISSUE TITLE
Implement the presence module

### ISSUE THREAD

**Issue by babolivier** [matrix-org/dendrite#602]
- [ ] Implement the [m.presence](https://matrix.org/docs/spec/client_server/r0.4.0.html#m-presence) event 
- [ ] Implement [PUT /_matrix/client/r0/presence/{userId}/status](https://matrix.org/docs/spec/client_server/r0.4.0.html#put-matrix-client-r0-presence-userid-status)
- [ ] Implement [GET /_matrix/client/r0/presence/{userId}/status](https://matrix.org/docs/spec/client_server/r0.4.0.html#get-matrix-client-r0-presence-userid-status)
- [ ] Implement [POST /_matrix/client/r0/presence/list/{userId}](https://matrix.org/docs/spec/client_server/r0.4.0.html#post-matrix-client-r0-presence-list-userid)
- [ ] Implement [GET /_matrix/client/r0/presence/list/{userId}](https://matrix.org/docs/spec/client_server/r0.4.0.html#get-matrix-client-r0-presence-list-userid)
- [ ] Implement setting and getting presence through [/sync](https://matrix.org/docs/spec/client_server/r0.4.0.html#get-matrix-client-r0-sync)

**Comment by Kegsay** [comment]
Related sytests:
```
    × GET /presence/:user_id/status fetches initial status
    × PUT /presence/:user_id/status updates my presence
    × Presence change reports an event to myself
    × Friends presence changes reports events
    × Presence changes are reported to local room members
    × Presence changes are also reported to remote room members
    × Presence changes to UNAVAILABLE are reported to local room members
    × Presence changes to UNAVAILABLE are reported to remote room members
    × New federated private chats get full presence information (SYN-115)
    × Left room members do not cause problems for presence
    Sync API:
    × Existing members see new member's presence
    × Get presence for newly joined members in incremental sync
    × User sees their own presence in a sync
    × User sees updates to presence from other users in the incremental sync.
```

**Comment by alexkursell** [comment]
I would love to take a crack at implementing this! I assume that the implementation would be somewhat similar to #1528.

**Comment by ptman** [comment]
Assuming this is the resource sink it is for synapse, there should be a toggle for turning presence off.

**Comment by S7evinK** [comment]
> I would love to take a crack at implementing this! I assume that the implementation would be somewhat similar to #1528.

@alexkursell: Are you still on it or need help? Should be pretty similar to #1528.

**Comment by neilalexander** [comment]
This landed in Dendrite 0.8.0.
